File: src/main/webapp/js/app/RecipeService.js
app.factory('recipeService', function RecipeService($http){

    const recipesPath = "/recipes/";

    return {
        getRootRecipes :function () {
            return $http.get(recipesPath);
        },

        getRelatedRecipes :function (id) {
            return $http.get(recipesPath + id + "/related");
        },

        getRecipe :function (id, version) {
            let params = version ? { version: version } : {}
            return $http.get(recipesPath + id, { params: params });
        },

        createRecipe :function (recipe) {
            return $http.post(recipesPath, recipe);
        },

        updateRecipe :function (recipe) {
            return $http.put(recipesPath + recipe.id, recipe);
        }
    }
})

File: src/main/java/com/cookbook/service/RecipeService.java
package com.cookbook.service;

import com.cookbook.model.Recipe;

import java.util.List;

public interface RecipeService {

    List<Recipe> getRootRecipes();

    List<Recipe> getRelatedRecipes(long id);

    Recipe getRecipe(long id);

    Recipe getRecipe(long id, int version);

    Recipe createRecipe(Recipe recipe);

    Recipe updateRecipe(Recipe recipe, long id);
}
